File: Data Structures & Algorithms/two-integer-sum/submission-2.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        int p1 = 0, p2 = nums.size()-1;
        map<int, vector<int>> idx;
        for(int i = 0; i < nums.size(); i++){
            idx[nums[i]].push_back(i);
        }
        sort(nums.begin(),nums.end());
        vector<int> res;
        while(p1 < p2){
            if(nums[p1]+nums[p2] == target){
                res.push_back(idx[nums[p1]][0]);
                if(nums[p1] == nums[p2])
                    res.push_back(idx[nums[p2]][1]);
                else
                    res.push_back(idx[nums[p2]][0]);
                sort(res.begin(), res.end());
                return res;
            }
            else if(nums[p1]+nums[p2] > target){
                p2--;
            }
            else
                p1++;
        }

        return res;
    }
};

File: Data Structures & Algorithms/car-fleet/submission-2.java
class Solution {
    public int carFleet(int target, int[] position, int[] speed) {
        int len = position.length;
        int[][] pair = new int[len][2];
        Stack<Double> st = new Stack<>();

        for(int i = 0; i < len; i++){
            pair[i][0] = position[i];
            pair[i][1] = speed[i];
        }

        Arrays.sort(pair, (a,b) -> Integer.compare(b[0], a[0]));

        for(int i = 0; i < len; i++){
            double res = (double)(target - pair[i][0])/pair[i][1];
            if(st.empty())
                st.push(res);
            else if(!st.empty() && res > st.peek())
                st.push(res);
        }

        return st.size();
    }
}
